airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/HeartbeatTimeoutChaperone.java

```diff
@@ -0,0 +1,149 @@
+  private final Optional<Runnable> customMonitor;
+
+  private boolean hasFailed = false;
+
+  public HeartbeatTimeoutChaperone(final HeartbeatMonitor heartbeatMonitor,
+                                   final Duration timeoutCheckDuration,
+                                   final FeatureFlagClient featureFlagClient,
+                                   final UUID workspaceId) {
+    this.timeoutCheckDuration = timeoutCheckDuration;
+
+    this.heartbeatMonitor = heartbeatMonitor;
+    this.featureFlagClient = featureFlagClient;
+    this.workspaceId = workspaceId;
+    this.customMonitor = Optional.empty();
+  }
+
+  @VisibleForTesting
+  public HeartbeatTimeoutChaperone(final HeartbeatMonitor heartbeatMonitor,
+                                   final Duration timeoutCheckDuration,
+                                   final FeatureFlagClient featureFlagClient,
+                                   final UUID workspaceId,
+                                   final Optional<Runnable> customMonitor) {
+    this.timeoutCheckDuration = timeoutCheckDuration;
+
+    this.heartbeatMonitor = heartbeatMonitor;
+    this.featureFlagClient = featureFlagClient;
+    this.workspaceId = workspaceId;
+    this.customMonitor = customMonitor;
+  }
+
+  public Runnable runWithHeartbeatThread(final Runnable runnable) {
+    LOGGER.info("Starting source heartbeat check. Will check every {} minutes.", timeoutCheckDuration.toMinutes());
+    return () -> {
+      final CompletableFuture<Void> runnableFuture = CompletableFuture.runAsync(runnable, executorService);
+      final CompletableFuture<Void> heartbeatFuture = CompletableFuture.runAsync(customMonitor.orElse(this::monitor), executorService);
+
+      try {
+        CompletableFuture.anyOf(runnableFuture, heartbeatFuture).get();
+      } catch (final InterruptedException e) {
+        LOGGER.error("Heartbeat chaperone thread was interrupted.", e);
+        return;
+      } catch (final ExecutionException e) {
+        // this should check explicitly for source and destination exceptions
+        if (e.getCause() instanceof RuntimeException) {
+          throw (RuntimeException) e.getCause();
+        } else {
+          throw new RuntimeException(e);
+        }
+      }
+
+      LOGGER.info("thread status... heartbeat thread: {} , replication thread: {}", heartbeatFuture.isDone(), runnableFuture.isDone());
+
+      boolean heartbeatTimedOut = false;
+      if (heartbeatFuture.isDone() && !runnableFuture.isDone()) {
+        heartbeatTimedOut = true;
+        runnableFuture.cancel(true);
+      }
+
+      if (runnableFuture.isDone() && !heartbeatFuture.isDone()) {
+        heartbeatFuture.cancel(true);
+      }
+
+      if (heartbeatTimedOut) {
+        throw new HeartbeatTimeoutException(String.format("Heartbeat has stopped. Heartbeat freshness threshold: %s Actual heartbeat age: %s",
+            heartbeatMonitor.getHeartbeatFreshnessThreshold(),
```

Comment:
I think that we will want some additional formatting on these `Duration` objects here, as the toString of duration uses the ISO-8601 notation (e.g. `PT8H6M12.345S`).  I'm not sure how readable that is to end users, especially non-technical ones.  I think it would probably make more sense to standardize on a unit of measure (seconds, perhaps) and just log the number of seconds.

Reply:
Done (the exception creation has been moved)